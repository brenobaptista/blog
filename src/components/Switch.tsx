import { useContext } from 'react'

import ThemeContext from '@/contexts/ThemeContext'
import {
  InvisibleCheckbox,
  Background,
  Toggle
} from '@/styles/components/Switch'

const Switch = (): JSX.Element => {
  const { theme, toggleTheme } = useContext(ThemeContext)

  const handleKeyPress = (event: React.KeyboardEvent<HTMLLabelElement>) => {
    if (event.key !== ' ') return

    event.preventDefault()
    toggleTheme()
  }

  return (
    <>
      <InvisibleCheckbox
        id='checkbox'
        type='checkbox'
        checked={theme.mode === 'dracula'}
        onChange={toggleTheme}
      />
      <Background
        htmlFor='checkbox'
        tabIndex={0}
        onKeyDown={event => handleKeyPress(event)}
        data-testid='label-checkbox'
        aria-label='Change the visual styles'
      >
        <Toggle />
      </Background>
    </>
  )
}

export default Switch
