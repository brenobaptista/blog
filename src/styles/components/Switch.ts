import styled from 'styled-components'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 2px;
`

export const Label = styled.div`
  font-size: 0.6rem;
  font-weight: 500;
`

export const InvisibleCheckbox = styled.input`
  position: absolute;
  height: 0;
  width: 0;
  visibility: hidden;
`

export const Background = styled.label`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: relative;
  cursor: ${props =>
    props.theme.mode === 'retro'
      ? "url('/images/retro/cursor/pointer.cur'), pointer"
      : 'pointer'};
  border-radius: 56px;
  width: 56px;
  height: 28px;
  background-color: ${props => props.theme.colors.visited};
`

export const Toggle = styled.span`
  content: '';
  position: absolute;
  top: 2px;
  left: 2px;
  border-radius: 24px;
  width: 24px;
  height: 24px;
  background-color: ${props => props.theme.colors.background};
  transition: 0.2s all ease-in-out;

  ${InvisibleCheckbox}:checked + ${Background} & {
    left: calc(100% - 2px);
    transform: translateX(-100%);
  }
`
