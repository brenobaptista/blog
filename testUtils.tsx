import { render } from '@testing-library/react'
import { RenderResult } from '@testing-library/react/types'
import { ThemeProvider } from 'styled-components'

import light from './src/styles/themes/light'

const Providers = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
)

const customRender = (ui: React.ReactElement, options = {}): RenderResult =>
  render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'

export { customRender as render }
