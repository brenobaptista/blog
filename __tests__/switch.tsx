import { render, fireEvent } from '../testUtils'
import Switch from '../src/components/Switch'

test('matches snapshot', () => {
  const toggleSwitch = jest.fn()
  const { asFragment } = render(
    <Switch switchValue toggleSwitch={toggleSwitch} />
  )

  expect(asFragment()).toMatchSnapshot()
})

test('toggles Switch using keyboard', () => {
  const toggleSwitch = jest.fn()
  const { getByTestId } = render(
    <Switch switchValue toggleSwitch={toggleSwitch} />
  )

  const checkbox = getByTestId('label-checkbox')
  fireEvent.keyDown(checkbox, { key: 'Enter', keyCode: 13 })
  fireEvent.keyDown(checkbox, { key: ' ', keyCode: 32 })
  expect(toggleSwitch).toHaveBeenCalled()
})
