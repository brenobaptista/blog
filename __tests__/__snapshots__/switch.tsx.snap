// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c1 {
  position: absolute;
  height: 0;
  width: 0;
  visibility: hidden;
}

.c0:checked + .c2 .c3 {
  left: calc(100% - 2px);
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

<input
    checked=""
    class="c0 c1"
    id="checkbox"
    type="checkbox"
  />
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  margin-bottom: 32px;
  cursor: pointer;
  border-radius: 65px;
  width: 65px;
  height: 32px;
  background: #586693;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
}

.c3 {
  content: '';
  position: absolute;
  top: 2px;
  left: 2px;
  border-radius: 28px;
  width: 28px;
  height: 28px;
  -webkit-transition: 0.2s;
  transition: 0.2s;
  background: #f6f5f4;
  box-shadow: 0 0 2px 0 rgba(10,10,10,0.29);
}

.c0:active .c2 {
  width: 33px;
}

.c4:checked + .c0 .c2 {
  left: calc(100% - 2px);
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

<label
    class="c0 c1"
    data-testid="label-checkbox"
    for="checkbox"
    tabindex="0"
  >
    <span
      class="c2 c3"
    />
  </label>
</DocumentFragment>
`;
